File: cpp/310_find_min_height_trees/find_min_height_trees.cpp
/*
 * @Date: 2022-04-06 11:08:29
 * @Author: Mengsen Wang
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-03-20
 * @FilePath: /algorithm/cpp/310_find_min_height_trees/find_min_height_trees.cpp
 */

#include <cassert>
#include <queue>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> findMinHeightTrees(int n, vector<vector<int>> edges) {
    if (n == 1) {
      return {0};
    }
    vector<int> degree(n);
    vector<vector<int>> adj(n);
    for (auto& edge : edges) {
      adj[edge[0]].emplace_back(edge[1]);
      adj[edge[1]].emplace_back(edge[0]);
      degree[edge[0]]++;
      degree[edge[1]]++;
    }
    queue<int> qu;
    vector<int> ans;
    for (int i = 0; i < n; i++) {
      if (degree[i] == 1) {
        qu.emplace(i);
      }
    }
    int remainNodes = n;
    while (remainNodes > 2) {
      int sz = qu.size();
      remainNodes -= sz;
      for (int i = 0; i < sz; i++) {
        int curr = qu.front();
        qu.pop();
        for (auto& v : adj[curr]) {
          if (--degree[v] == 1) {
            qu.emplace(v);
          }
        }
      }
    }
    while (!qu.empty()) {
      ans.emplace_back(qu.front());
      qu.pop();
    }
    return ans;
  }
};

int main() {
  vector<tuple<int, vector<vector<int>>, vector<int>>> tests{
      {4, {{1, 0}, {1, 2}, {1, 3}}, {1}},
      {6, {{3, 0}, {3, 1}, {3, 2}, {3, 4}, {5, 4}}, {3, 4}},
  };

  for (auto& [n, edges, ans] : tests) {
    assert(Solution().findMinHeightTrees(n, edges) == ans);
  }
}

File: golang/310_find_min_height_trees/find_min_height_trees.go
/*
 * @Date: 2022-04-06 11:08:38
 * @Author: Mengsen Wang
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-03-20
 * @FilePath: /algorithm/golang/310_find_min_height_trees/find_min_height_trees.go
 */

package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func findMinHeightTrees(n int, edges [][]int) []int {
	if n == 1 {
		return []int{0}
	}

	g := make([][]int, n)
	deg := make([]int, n)
	for _, e := range edges {
		x, y := e[0], e[1]
		g[x] = append(g[x], y)
		g[y] = append(g[y], x)
		deg[x]++
		deg[y]++
	}

	q := []int{}
	for i, d := range deg {
		if d == 1 {
			q = append(q, i)
		}
	}

	remainNodes := n
	for remainNodes > 2 {
		remainNodes -= len(q)
		tmp := q
		q = nil
		for _, x := range tmp {
			for _, y := range g[x] {
				deg[y]--
				if deg[y] == 1 {
					q = append(q, y)
				}
			}
		}
	}
	return q
}

func main() {
	tests := []struct {
		n     int
		edges [][]int
		ans   []int
	}{
		{4, [][]int{{1, 0}, {1, 2}, {1, 3}}, []int{1}},
		{6, [][]int{{3, 0}, {3, 1}, {3, 2}, {3, 4}, {5, 4}}, []int{3, 4}},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, findMinHeightTrees(test.n, test.edges), index)
	}
}
